[improvement](orc-reader) Implements ORC lazy materialization

kaka11chen opened the pull request:
# Proposed changes


## Problem summary

- Implements ORC lazy materialization, integrate with the implementation of https://github.com/apache/doris-thirdparty/pull/56 and https://github.com/apache/doris-thirdparty/pull/62.
- Refactor code: Move `execute_conjuncts()` and `execute_conjuncts_and_filter_block()` in `parquet_group_reader `to `VExprContext`, used by parquet reader and orc reader.
- Add session variables `enable_parquet_lazy_materialization` and `enable_orc_lazy_materialization` to control whether enable lazy materialization.
- Modify `build.sh` to update apache-orc submodule or download package every time.
 
## Performance test results

- 100 million rows of random data.
- SQL: `select type_col1, type_col2 from tables where filter`. The filter will returns 1400 records.
- Single thread test.

  | orc-snappy-lazy | orc-snappy-no-lazy
-- | -- | --
tinyint | 2.26 | 2.83
smallint | 2.30 | 2.86
int | 2.07 | 2.32
bigint | 2.38 | 3.21
boolean | 2.29 | 3.15
float | 2.30 | 3.34
double | 2.55 | 3.69
string | 3.97 | 11.22
binary | 3.79 | 10.28
timestamp | 2.46 | 24.38
decimal | 2.61 | 7.00
char | 2.71 | 8.29
varchar | 2.87 | 5.80
date | 2.36 | 21.82

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

morningman commented on build.sh:
```diff
@@ -247,16 +247,13 @@ if [[ ! -f "${DORIS_THIRDPARTY}/installed/lib/libbacktrace.a" ]]; then
 fi
 
-if [[ ! -f "${DORIS_HOME}/be/src/apache-orc/README.md" ]]; then
+if [[ ! -f "${DORIS_HOME}/be/src/apache-orc/README.md " ]]; then
```
```suggestion
if [[ ! -f "${DORIS_HOME}/be/src/apache-orc/README.md" ]]; then
```

morningman reviewed (dismissed):
LGTM

kaka11chen commented:
run buildall

hello-stephen commented:
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.2 seconds
 stream load tsv:          426 seconds loaded 74807831229 Bytes, about 167 MB/s
 stream load json:         23 seconds loaded 2358488459 Bytes, about 97 MB/s
 stream load orc:          59 seconds loaded 1101869774 Bytes, about 17 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230427154035_clickbench_pr_136710.html

kaka11chen committed:
[Feature](orc-reader) Implements ORC lazy materialization.

- Implements ORC lazy materialization, integrate with the implementation of https://github.com/apache/doris-thirdparty/pull/56 and https://github.com/apache/doris-thirdparty/pull/62.
- Refactor code: Move `execute_conjuncts()` and `execute_conjuncts_and_filter_block()` in `parquet_group_reader `to `VExprContext`, used by parquet reader and orc reader.
- Add session variable `enable_parquet_lazy_materialization` and `enable_orc_lazy_materialization` to control whether enable lazy materialization.

kaka11chen committed:
Fix compile issue.

kaka11chen committed:
Update build.sh.

kaka11chen commented:
run buildall

kaka11chen committed:
Fix _init_read_columns() location.

kaka11chen commented:
run buildall

morningman reviewed (approved):
LGTM
